diff --git a/clvk/src/memory.cpp b/clvk/src/memory.cpp
index a655c64..7089c4f 100644
--- a/clvk/src/memory.cpp
+++ b/clvk/src/memory.cpp
@@ -263,9 +263,7 @@ cvk_image::required_format_feature_flags_for(cl_mem_object_type type,
     if (type == CL_MEM_OBJECT_IMAGE1D_BUFFER) {
         format_feature_flags_RO = VK_FORMAT_FEATURE_UNIFORM_TEXEL_BUFFER_BIT;
     } else {
-        format_feature_flags_RO =
-            VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT |
-            VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT;
+        format_feature_flags_RO = VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT;
     }
     VkFormatFeatureFlags format_feature_flags_WO;
     if (type == CL_MEM_OBJECT_IMAGE1D_BUFFER) {
