diff --git a/backend/epsonds.c b/backend/epsonds.c
index 72d01d8..de79826 100644
--- a/backend/epsonds.c
+++ b/backend/epsonds.c
@@ -1259,6 +1259,9 @@ const  epsonds_profile_map epsonds_models_predefined[] = {
   {0x118A, "PID 118A","L3250 Series", 7},
   {0x119B, "PID 119B","XP-2150 Series", 7},
   {0x11B1, "PID 11B1","XP-2200 Series", 7},
+  {0x0193, "ES-C220","ES-C220", 5},
+  {0x018F, "DS-C330","DS-C330", 5},
+  {0x0191, "DS-C490","DS-C490", 5},
   {0x00, "","", 0x00 }
 };
 
diff --git a/doc/descriptions/epsonds.desc b/doc/descriptions/epsonds.desc
index 989caf8..8e74c50 100644
--- a/doc/descriptions/epsonds.desc
+++ b/doc/descriptions/epsonds.desc
@@ -358,3 +358,18 @@
 :interface  "USB"
 :usbid      "0x04b8" "0x11b1"
 :status     :complete
+
+:model      "ES-C220"
+:interface  "USB"
+:usbid      "0x04b8" "0x0193"
+:status     :complete
+
+:model      "DS-C330"
+:interface  "USB"
+:usbid      "0x04b8" "0x018f"
+:status     :complete
+
+:model      "DS-C490"
+:interface  "USB"
+:usbid      "0x04b8" "0x0191"
+:status     :complete
