From a2452a9bc2a582e9da2deaaa4deac7fc38ebb14e Mon Sep 17 00:00:00 2001
From: Steven 'Steve' Kendall <skend@google.com>
Date: Thu, 20 Jul 2023 20:58:11 +0000
Subject: [PATCH] enable V1 configurations

---
 ucm2/ucm.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ucm2/ucm.conf b/ucm2/ucm.conf
index fb1a0be..abb4bc2 100644
--- a/ucm2/ucm.conf
+++ b/ucm2/ucm.conf
@@ -13,7 +13,7 @@
 
 Syntax 4
 
-Define.V1 ""		# non-empty string to enable ucm v1 paths
+Define.V1 yes		# non-empty string to enable ucm v1 paths
 Define.V2ConfD yes	# empty string to disable
 Define.V2Module ""	# non-empty string to enable module name lookups (obsolete)
 Define.V2Name ""	# non-empty string to enable driver & card name lookups (obsolete)
-- 
2.41.0.487.g6d72f3e995-goog

